cpp/src/arrow/dataset/scanner.cc

```diff
@@ -877,6 +882,15 @@ Result<compute::ExecNode*> MakeScanNode(compute::ExecPlan* plan,
         batch->values.emplace_back(partial.fragment.index);
         batch->values.emplace_back(partial.record_batch.index);
         batch->values.emplace_back(partial.record_batch.last);
+
+        auto filefragment_casted =
+            dynamic_cast<const FileFragment*>(partial.fragment.value.get());
```

Comment:
Use `arrow::internal::checked_cast`? This will be dynamic_cast in debug mode and static cast otherwise

Reply:
Made the change.

Reply:
I am not very sure for the static_cast. We basically need to check whether it is actually an instance of the FileFragment and if it is, only then have to include the filename.